### PR TITLE
Fixes converted ads keep showing

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/9436

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

See https://github.com/brave/brave-browser/issues/9436

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 1]
```diff
@@ -1094,6 +1095,10 @@ std::vector<std::unique_ptr<ExclusionRule>>
       std::make_unique<TotalMaxFrequencyCap>(frequency_capping_.get());
   exclusion_rules.push_back(std::move(total_max_frequency_cap));
 
+  std::unique_ptr<ExclusionRule> conversion_frequency_cap =
```
Comment (NejcZdovc) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:1098]: nit: could be auto

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc [hunk 2]
```diff
@@ -1094,6 +1095,10 @@ std::vector<std::unique_ptr<ExclusionRule>>
...
+      std::make_unique<ConversionFrequencyCap>(frequency_capping_.get());
```
Comment (NejcZdovc) [inline vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:1099]: `frequency_capping_` null check is missing

File: vendor/bat-native-ads/src/bat/ads/internal/client_mock.h [hunk 3]
```diff
@@ -38,6 +38,11 @@ class ClientMock : public Client {
       const std::string& segment,
       const uint64_t time_offset_in_seconds,
       const uint16_t weight);
+
+  void GeneratePastAdConversionHistoryFromNow(
+      const std::string& creative_set_id,
+      const int64_t time_offset_per_ad_in_seconds,
+      const uint8_t count);
```
Comment (NejcZdovc) [inline vendor/bat-native-ads/src/bat/ads/internal/client_mock.h:45]: `#include <stdint.h>` is missing

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#include "bat/ads/internal/client.h"
+#include "bat/ads/internal/frequency_capping/frequency_capping.h"
+
+#include "base/logging.h"
+#include "base/strings/stringprintf.h"
+
+namespace ads {
+
+ConversionFrequencyCap::ConversionFrequencyCap(
+    const FrequencyCapping* const frequency_capping)
```
Comment (NejcZdovc) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc:18]: double const?

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+ConversionFrequencyCap::~ConversionFrequencyCap() = default;
+
+bool ConversionFrequencyCap::ShouldExclude(
+    const CreativeAdInfo& ad) {
+  if (!DoesRespectCap(ad)) {
+    last_message_ = base::StringPrintf("creativeSetId %s has exceeded the "
+        "frequency capping for conversions", ad.creative_set_id.c_str());
+
```
Comment (NejcZdovc) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc:30]: nit: extra line

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc:30]: intentional :-)

File: vendor/bat-native-ads/src/bat/ads/internal/client_mock.h (hunk 3 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/client_mock.h:45]: This has been fixed as part of a wider refactor of unit tests/mocks

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc (hunk 2 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:1099]: This has been fixed as part of a wider refactor of unit tests/mocks

File: vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc (hunk 1 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/ads_impl.cc:1098]: I only use `auto` now for complex types as makes code harder to read

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc (hunk 4 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc:18]: const X* const p means “p is a const pointer to an X that is const”: you can’t change the pointer p itself, nor can you change the X object via p.

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/exclusion_rules/conversion_frequency_cap.cc:18]: @NejcZdovc see https://isocpp.org/wiki/faq/const-correctness#const-ptr-vs-ptr-const

**Comment by btlechowski** [comment]
LGTM

Brave | 1.10.28 Chromium: 81.0.4044.129 (Official Build) nightly (64-bit)
-- | --
Revision | 3d71af9f5704a40b85806f4d08925db24605ba25-refs/branch-heads/4044@{#979}
OS | Linux

Verified test plan from https://github.com/brave/brave-browser/issues/9436.
Also checked the logs for `creativeSetId (...) has exceeded the frequency capping for conversions` message.
Verified ads from the same creativeSetId were not shown after conversion.

```
[6080:1:0505/010858.182044:INFO:ads_impl.cc(359)] Browser state changed to unidle
[6080:1:0505/010858.204730:INFO:ads_impl.cc(954)] Serving ad from categories:
[6080:1:0505/010858.204876:INFO:ads_impl.cc(956)]   technology & computing-technology & computing
[6080:1:0505/010858.205649:INFO:ads_impl.cc(956)]   arts & entertainment-video games
[6080:1:0505/010858.205814:INFO:ads_impl.cc(956)]   technology & computing-software
[6080:1:0505/010858.207833:INFO:ads_impl.cc(974)] No eligible ads found in categories:
[6080:1:0505/010858.210006:INFO:ads_impl.cc(976)]   technology & computing-technology & computing
[6080:1:0505/010858.216231:INFO:ads_impl.cc(976)]   arts & entertainment-video games
[6080:1:0505/010858.216346:INFO:ads_impl.cc(976)]   technology & computing-software
[6080:1:0505/010858.216469:INFO:ads_impl.cc(1007)] Serving ad from parent categories:
[6080:1:0505/010858.216551:INFO:ads_impl.cc(1009)]   technology & computing
[6080:1:0505/010858.216631:INFO:ads_impl.cc(1009)]   arts & entertainment
[6080:1:0505/010858.231964:INFO:ads_impl.cc(1188)] All advertisers have been shown, so round robin
[6080:1:0505/010858.234288:INFO:client.cc(420)] Resetting seen advertisers
[6080:1:0505/010858.234825:INFO:ads_impl.cc(1130)] creativeSetId de5a82e1-17d6-47e0-a368-17f8f56dfeb3 has exceeded the frequency capping for conversions
[6080:1:0505/010858.235719:INFO:ads_impl.cc(974)] No eligible ads found in categories:
[6080:1:0505/010858.235923:INFO:ads_impl.cc(976)]   technology & computing
[6080:1:0505/010858.236101:INFO:ads_impl.cc(976)]   arts & entertainment
[6080:1:0505/010858.236275:INFO:ads_impl.cc(1020)] Serving ad notification from untargeted category
[6080:1:0505/010858.243086:INFO:client.cc(593)] Successfully saved client state
[6080:1:0505/010858.243522:INFO:ads_impl.cc(1080)] Notification not made: No eligible ads found
```

Verified that without conversion, ads keep showing for the same creative set

**Pull request merged**